doc/tgl.1
.Dd 16 June 2012
.Dt TGL 1
.Os
.Sh NAME
.Nm tgl
.Nd Run the Text Generation Language interpreter
.Sh SYNOPSIS
.Nm
.Op Fl h
.Op Fl p
.Op Fl c Ar context
.Op Fl l Ar library
.Op Fl r Ar file
.Sh DESCRIPTION
Runs the Text Generation Language interpreter on the script read from standard
input.
.Pp
tgl supports the following options:
.Bl -tag -width Ds
.It Fl h
Show usage and exit.
.It Fl p
Search for prefix payload data.
.It Fl c Ar context
Specify the current context (for conditional execution).
.It Fl l Ar library
Use the file specified by
.Ar library
(instead of ~/.tgl) for the user library.
.It Fl r Ar file
Use
.Ar file
(instead of ~/.tgl_registers) for register persistence.
.El
.Pp
When Tgl starts up, it first restores registers from the register persistence
file (by default, ~/.tgl_registers), then executes the user library if it
exists (by default, ~/.tgl). Any errors in either of these steps are reported
but otherwise ignored.
.Pp
The script to execute is then read from standard input, until EOF is
reached. The script is then executed; if any error occurs, execution is
aborted. If no errors are encountered, the final state of the registers is
written back to the register persistence file.
.Sh EXIT STATUS
.Nm
exits with status 0 on success. An exit status of 1 indicates failure due
to an error in the input program. Stati 253--255 indicate failures due to
external conditions.
.Sh BASIC CONCEPTS
.Ss VALUES
Every value in Tgl is a string. A string is simply a sequence of arbitrary
8-bit bytes of arbitrary length. Some commands will interpret the strings as
integers or booleans. An integer is a string beginning with a
.Qq +
or
.Qq - ,
followed by any number of base-10 digits. Base-16 (hexadecimal) may be
specified by prefixing the digits with
.Qq 0x ,
base-8 (octal) by
.Qq 0o ,
or base-2 (binary) by
.Qq 0b .
A boolean string is false if it is the empty string or if it is a valid
representation of numeric zero. Any other string is considered true.
.Ss STACK
The stack is the primary method of passing data to and from commands. With few
exceptions, all commands only pop values from and push values to the top of the
stack. There is no limit on stack height, other than available memory. It is an
error to try to pop from the stack when it is empty.
.Ss REGISTERS
Tgl provides the programmer with exactly 256 registers, each corresponding to a
single byte value. Each register contains one string, and also stores its time
of last access. Other than the history system, commands do not use registers
except when doing so to present functionality to the user.
.Pp
Registers are all initialised to empty strings. When Tgl exits successfully, it
will write the state of all registers to the register persistence file. This
file is read back in the next time Tgl runs, so registers will preserve their
values across invocations of Tgl.
.Pp
Registers 0x00..0x1F (ie, the byte values) are used to store command history,
where register 0 is the most recently run command sequence. When TGL runs
successfully and history was not suppressed, and the command sequence was not
\(dqhX\(dq, the command sequence is added to the history by moving 0x00..0x1E
each one register up, then inserted to register 0x00.
.Ss SECONDARY ARGUMENTS
Some commands take optional parameters via a secondary argument system. There
are four secondary argument slots, called U0..U3. Unlike registers, they cannot
be accessed directly, and may be null. Any command which takes secondary
arguments nullifies all four slots after execution. There is also the secondary
index, UX, which indicates which slot is used next. It starts at zero, and is
incremented whenever a secondary argument is set. It is zeroed when a
secondary-argument-using command completes execution.
.Pp
A secondary argument is considered
.Qq set
when UX has advanced beyond its own index. It is considered
.Qq provided
when it is set and non-null.
.Pp
Commands which use secondary arguments indicate such usage similar to the
example below:
.Dl Secondary: [value = default] [provided] (set?)
.Pp
Brackets indicate that the argument is considered if provided; parantheses, if
set. Items of the format value=default indicate a property which has the given
default if the argument is not provided, but which takes the value of any
argument which is provided. Items of the format property? indicate booleans
which are false unless the argument is provided or set (depending on whether
brackets or parentheses are used), regardless of the argument's value. Any
argument not noted is silently ignored.
.Pp
The
.Li u
(secondary-argument) command is used to set secondary arguments. What it
does depends on the following character:
.Bl -tag -width Ds
.It %
Pop a value from the main stack and use it to provide the next argument
.It space
Set the next argument to null
.It digit
Evaluate the following digits (using the number command) and use the result to
provide the next argument
.It "- +"
Same as digit
.It .
Provides an integer corresponding to the current stack height as the next
argument.
.It other
Create a one-character string with the given character and use it to provide
the next argument
.El
.Ss PAYLOAD DATA
In some cases, it is desirable to take data from a structured source, either
embedded in the script itself, or extracted from an external source. The
.Li ,
command provides functionality to access and manipulate this payload data.
.Pp
Most commonly, one includes payload data within the script itself. Usually,
payload data starts after the text
.Qq ",$" ,
though the delimiter can be changed. By passing the
.Li -p
(or
.Li --prefix-payload )
flag, it is also possible to include a payload at the beginning of input. In
this case, code is considered to start after the first
.Qq "|"
(though initial whitespace (for the
.Li "`"
character within strings) still counts within the payload). Multiple pipes can
be used to include the
.Qq "|"
within the payload; the first occurance that has the greatest number of
pipes is used to separate the payload and the code.
.Pp
The first occurrance of a payload-data command which requires access to the
payload will search for the start-of-payload delimiter (default
.Qq ",$" )
in the code if no payload is already present.
.Ss COMMANDS
All
.Qq true commands
in TGL are represented by a single character. Whenever the interpreter must
execute a command, it looks at the character at the interpreter's current
instruction pointer, and runs the associated command. All whitespace characters
are implicitly considered no-op commands. If the command completes
successfully, the instruction pointer is incremented.
.Pp
Builtin commands have access to all interpreter state, including the code and
instruction pointer itself. Many commands will move the instruction pointer to
effect control and quoting mechanisms. See the description for each command.
.Ss CONTEXT
TGL is run with a context (the empty string if not given) and maintains whether
the context is active. The context is typically specified as a filename. Some
commands only have effect if the context is active, allowing for language- or
project-specific functionality.
.Sh CATEGORICAL COMMAND REFERENCE
.Ss STACK
.Bl -tag -width Ds
.It x (swap: a{depth} b -> b a{depth}, or b a{-depth} -> a{-depth} b)
.Dl Secondary: [depth = 1]
The top two items on the stack are swapped. With a secondary argument, this
rolls the top element under that many items if positive, or moves an item at
that depth to the top if negative.
.It ":" (dupe: item -> item{times+1})
.Dl Secondary: [times = 1]
Pops an item from the stack, then pushes (times+1) copies of it onto the stack.
.It ";" (drop: item{times} -> ())
.Dl Secondary: [times = 1]
Pops (times) items from the stack and discards them.
.El
.Ss STRINGS
.Bl -tag -width Ds
.It c (concat: a b -> ab)
Pops two strings from the stack, concatenates them, and pushes the result.
.It C (char-at: string index -> char)
Pops a string and an index, and pushes the character within the string at that
index. The first character is at index 0. Negative indices are relative to the
end of the string.
.It l (length: string -> length)
Pops a string from the stack and pushes its length, in bytes.
.It m (map: string {from to}{n} n -> result, or string ... -> result)
.Dl Secondary: [stack-position-of-string = null]
Translates each occurrance of
.Ar from
to the corresponding
.Ar to
in
.Ar string
such that no mappings interact. The number of mappings must be specified
explicitly on the stack if no secondary argument is given; if one is given, it
should be a
.Li \(dqu.\(dq
command after
.Ar string
but before the first mapping.
.It s (substr: string from to -> length)
Pops a string and two indices, evaluated the same was as in char-at, and pushes
the portion of the string between
.Ar from ,
inclusive, to
.Ar to ,
exclusive.
.It S (suffix: string from -> length)
Same as substr, but assumes that
.Ar to
is the length of the string.
.It y (empty-string: () -> \(dq\(dq)
Pushes the empty string onto the stack.
.It \(dq ... \(dq (string: ??? -> string)
Pushes the characters between the double-quotes as a string onto the stack. The
following characters are interpreted specially:
.Bl -tag -width Ds
.It \e
Evaluates the
.Li \e
command at that point and splices the result into the string. Characters
continue to be read after the point where
.Li \e
left off.
.It "$"
Interprets the following character as a register name; the value of this
register is spliced into the string at that point.
.It "%"
Pops a value from the stack and splices it into the string at that point.
.It "`"
Inserts the string of whitespace characters that were present before the first
non-whitespace character in the primary input.
.El
.It "'" C (char: () -> string)
Pushes a single-character string, whose contents is the character following the
command, to the stack. Note that since TGL is not Unicode-aware, this more
strictly pushes a single
.Em byte .
For example, the code
.Dl "'ß"
will not push a string containing an Eszett, but rather (assuming UTF-8) a
string containing the byte 0xC3; the interpreter will then try to execute
command 0x9F.
.It "(" ... ")" (code: () -> string)
Pushes the string between the parentheses verbatim. The parentheses must be
balanced; for example, the code
.Dl "(foo (bar) baz)"
pushes the string
.Dl "foo (bar) baz"
.It "." (print: string -> ())
Pops a string from the stack and writes it verbatim to standard output.
.It \e (escape: () -> string or () -> ())
Pushes a single character as indicated by the following character(s).
.Pp
Control characters:
.Bl -tag -width Ds
.It \ea
BEL
.It \eb
BS
.It \ee
ESC
.It \ef
FF
.It \en
LF
.It \er
CR
.It \et
HT
.It \ev
VT
.It \ex##
The byte represented by the hexadecimal integer
.Ar ## .
.El
.Pp
The following result in the character being escaped:
.Dl \e\(dq \e\e \e$ \e% \e` \e'
.Pp
The following result in nothing (not even a string is pushed); they may be used
for manual balancing of parentheses as required by some commands.
.Dl \e( \e) \e[ \e] \e{ \e} \e< \e>
.It "=" (equal: a b -> (a==b))
Pops two strings and pushes a boolean indicating whether they are exactly
equal.
.It "!" (not-equal: a b -> (a!=b))
Pops two strings and pushes a boolean indicating whether they are
.Em not
exactly equal.
.It "{" (string-less: a b -> (a<b))
Pops two strings and pushes a boolean indicating whether the left is less than
(ASCIIbetically) the right.
.It "}" (string-greater: a b -> (a>b))
Pops two strings and pushes a boolean indicating whether the left is greater
than (ASCIIbetically) the right.
.El
.Ss MATH
.Bl -tag -width Ds
.It "#/digit" ... (number: () -> integer)
Reads an integer and pushes it onto the stack, stopping on the first
non-digit. Any digit (0..9) is bound to this command; the # command also allows
a leading sign before the number itself.
.It ? (rand: () -> integer)
Pushes a random integer between 0 and 65535, inclusive, onto the stack.
.It "+" (add: a b -> (a+b))
Pops two stack elements, interprets them as integers, and pushes their sum
onto the stack.
.It "-" (sub: a b -> (a-b))
Pops two stack elements, interprets them as integers, and pushes their
difference onto the stack.
.It "*" (mul: a b -> (a*b))
Pops two stack elements, interprets them as integers, and pushes their product
onto the stack.
.It "/" (div: a b -> (a/b))
Pops two stack elements, interprets them as integers, and pushes their quotient
onto the stack.
.It "%" (mod: a b -> (a%b))
Pops two stack elements, interprets them as integers, and pushes their
remainder onto the stack.
.It "<" (less: a b -> (a<b))
Pops two integers and pushes a boolean indicating whether the left is less than
the right.
.It ">" (greater: a b -> (a>b))
Pops two integers and pushes a boolean indicating whether the left is greater
than the left.
.It "&" (and: a b -> (a&&b))
Pops two booleans and pushes the result of logical AND. Note that due to the
nature of TGL this is not, and cannot be, short-circuiting.
.It "|" (or: a b -> (a||b))
Pops two booleans and pushes the result of logical OR. Note that due to the
nature of TGL this is not, and cannot be, short-circuiting.
.It "^" (xor: a b -> (a^b))
Pops two booleans and pushes the result of logical XOR.
.It "~" (not: a -> !a)
Pops a boolean and pushes its negation.
.El
.Ss REGISTERS
.Bl -tag -width Ds
.It a (auto-write: value -> ())
Pops a string and stores it into a register with an alphanumeric name. The
register chosen is the one with the earliest date of last use. After writing
the register, prints a string of the format
.Dl "`%s: %c\en"
indicating the value stored and the register it was stored in. (The first
character is a literal grave quote, while the last one is a linefeed.)
.It p (stash: () -> ())
Pushes all registers onto a stack independent of the primary stack.
.It P (retrieve: () -> ())
Restores all registers as preserved by stash, removing that entry.
.It r R (read: () -> value)
Pushes the contents of register
.Ar R
onto the stack.
.It R R (write: value -> ())
Pops
.Ar value
from the stack and stores it into register
.Ar R .
.It u X (secondary-argument: () -> () or value -> ())
The effect of this command depends on
.Ar X .
See SECONDARY ARGUMENTS above.
.It z (stash-retrieve: code -> guarded-code)
Pops
.Ar code
and places
.Li p...P
guards around it, so that no changes to registers occur due to execution of
.Ar code
(assuming it contains no \(dq,$\(dq).
.El
.Ss PAYLOAD
All payload commands are prefixed with \(dq,\(dq.
.Bl -tag -width Ds
.It ",$" (payload-start: () -> ())
Immediately moves the instruction pointer to the end of the current code,
effectively stopping processing.
.It ",c" (payload-curr: () -> elt)
Pushes the string corresponding to the first element of the current payload
data onto the stack.
.It ",," (payload-next: () -> ())
.Dl Secondary: [count = 1]
Deletes the first
.Ar count
elements from the current payload data.
.It ",;" (payload-next-kv: () -> ())
.Dl Secondary: [count = 1]
Deletes the first
.Ar count
pairs of elements from the current payload data.
.It ",." (payload-print: () -> ())
.Dl Secondary: [count = 1]
Prints the first
.Ar count
elements of the payload data, separated by the current
.Ar output-v-delimiter ,
then deletes them. A
.Ar count
of zero indicates all elements.
.It ",:" (payload-print-kv: () -> ())
.Dl Secondary: [count = 1]
Prints the first
.Ar count
key-value pairs of the payload data. Each key and value is separated by the
current
.Ar output-kv-delimiter ,
and each pair is separated by the current
.Ar output-kvs-delimiter .
A
.Ar count
of zero indicates all pairs.
.It ",r" (payload-read: () -> payload)
Pushes the string representation of the current payload data onto the stack.
.It ",R" (payload-write: payload -> ())
Sets the current payload data to the string popped from the stack.
.It ",x" (payload-recurse: payload code -> ())
Saves all current payload state, then sets the payload to
.Ar payload
and executes
.Ar code .
After execution, payload state (including delimiters and such) is reverted to
what is was before execution of this command.
.It ",h" (payload-length-bytes: () -> byte-count)
Pushes an integer indicating how many bytes remain in the current payload
data. Returns zero if and only if no elements remain in the current
payload. (Thus the mnemonic
.Em Has-more )
.It ",e" (payload-each: ??? body -> ???)
.Dl Secondary: [reg = \(dqp\(dq]
For each value in the current payload, set
.Ar reg
to that value and execute
.Ar body .
This does not modify the current payload data. The result of the payload data
having been altered after
.Ar body
exits is undefined.
.It ",E" (payload-each-kv: ??? body -> ???)
.Dl Secondary: [key-reg = \(dqk\(dq] [val-reg = \(dqv\(dq]
For each key-value pair in the current payload, set
.Ar key-reg
to the key and
.Ar val-reg
to the value, then execute
.Ar body .
This does not modify the current payload data. The result of the payload data
having been altered after
.Ar body
exits is undefined.
.It ",i" (payload-datum-at-index: index -> value)
Pushes the payload element at
.Ar index ,
where 0 is the index of the first element. This operation requires a linear
scan of the payload data.
.It ",I" (payload-num-indices: () -> count)
Counts the number of elements in the current payload data and pushes the
result. This operation requires a linear scan of the payload data.
.It ",k" (payload-datum-at-key: key -> value)
Pushes the first payload value element which follows the given key. This
opretation requires a linear scan of the payload data.
.It ",/" VV (payload-set-property: value -> ())
Sets the property
.Ar VV
to
.Ar value .
Properties are:
.Bl -hang -width Ds
.It ps (payload-start, default \(dq,$\(dq)
The delimiter that separates code from payload in the main body.
.It vd (value-delimiter, default \(dqws\(dq (see \(dq,s\(dq))
The delimiter that separates elements in the payload data.
.It ov (output-v-delimiter, default \(dq, \(dq)
The string to separate contiguous normal values on output.
.It ok (output-kv-delimiter, default \(dq, \(dq)
The string to separate keys from values on output.
.It os (output-kvs-delimiter, default \(dq\en\(dq)
The string to separate key-value pairs on output.
.It b( b[ b{ b< (balance-paren)
Set to any boolean. If true, delimiters are ignored when between yet-unbalanced
pairs of the given type of parenthesis character. If false, the characters are
treated normally. Default is true for ([{, false for <.
.It t( t[ t{ t< (trim-paren)
Set to any boolean. If true, up to one single matching pair of the given
parenthesis type is trimmed from a value before extraction. This is performed
after whitespace trimming, and no further trimming is done after this
step. Default is true for ([{, false for <.
.It ts (trim-space, default true)
Set to any boolean. If true, extraneous spaces around values are trimmed when
being extracted.
.El
.It ",?" VV (payload-get-property: () -> value)
Pushes the current value of property
.Ar VV .
See \(dq,/\(dq.
.It ",s" (payload-space-delimited: () -> ())
Sets payload to be delimited by whitespace of any kind, combination, and
abount. Resets balance and trim properties to defaults. If this mode is in use
when payload is first extracted (or with any command which alters the payload),
leading space in the payload is skipped. This is indicated in delimiter
properties with the special string \(dqws\(dq.
.It ",0" (payload-nul-delimited: () -> ())
Sets payload to be delimited by NUL characters, and sets all balance and trim
properties to false.
.It ",l" (payload-line-delimited: () -> ())
Sets the following strings:
.Dl \(dq\en\(dq \(dq\er\(dq \(dq\er\en\(dq
as the value delimiter, and sets all paren-related balancing and trimming to
false. Space trimming is set to true.
.It ",!" (payload-from-code: () -> ())
Extracts the payload from the suffix from the top-level primary code, using the
current
.Ar payload-start
delimiter. This is automatically done if no payload has ever been extracted
when a command that accesses the payload data is run.
.It ",f" (payload-from-file: filename -> ())
Reads the entire contents of
.Ar filename
and uses it as payload data.
.It ",F" (payload-from-glob: glob -> ())
Accumulates all filenames matching
.Ar glob
and stores them into the payload data, NUL-delimited. Implicitly executes
\(dq,0\(dq after.
.El
.Ss CONTROL STRUCTURES
.Bl -tag -width Ds
.It d (defun: name body -> ())
Defines the command of the given name, which may be standard (one character
long) or long (more than one character long). It is an error if it already
exists. After execution of this command, the new command can be invoked in the
normal manner (directly if a standard name, or with
.Li Q...
if it is long).
.It D (defun-contextual: name body -> ())
Like
.Li defun
except that nothing happens if the current context is not active.
.It e (each: ??? string body -> ???)
.Dl Secondary: [reg = \(dqc\(dq]
For each character in
.Ar string ,
set
.Ar reg
to a string with that character as its only content, then execute
.Ar body .
.It f (for: ??? to body -> ???)
.Dl Secondary: [from = 0] [reg = \(dqi\(dq] [increment = +1 or -1]
Set
.Ar reg
to
.Ar from .
Execute
.Ar body ,
then add
.Ar increment
to
.Ar reg .
If
.Ar reg
has not moved to the other side of
.Ar to
from whence it started and is not equal to
.Ar to ,
the command repeats. If
.Ar increment
is not given, it defaults to +1 if
.Ar to
is greater than
.Ar from ,
or -1 otherwise.
.It i (if: ??? condition then else -> ???)
If
.Ar condition
is a true boolean, executes
.Ar then ;
otherwise, executes
.Ar else .
Other than the three arguments given, the effect on the interpreter is entirely
dependent on the code executed. Note:
.Ar condition
is a
.Em value ,
not code to execute. The code
.Dl "(1 2=)('y)('n)i."
will print
.Li \(dqy\(dq
since
.Li \(dq1 2=\(dq
is a true string. For the desired effect, the code should read
.Dl "1 2=('y)('n)i."
.It I (if-short: ??? condition then -> ???)
The same as
.Li if
except that it does nothing if
.Ar condition
is false.
.It Q ... (long-command: ??? -> ???)
Reads characters from the code until end of input or whitespace is
encountered. These characters (excluding the leading
.Qq Q )
constitute a long command name. The command corresponding to this name is
executed. The way the interpreter is affected depends entirely on what the
command does.
.It v (save-code: name code -> ())
Creates a defun using the given name and code, executes it (the defun, not the
code), and appends the definition to the user library if successful.
.It V t (save-code-contextual: name code -> ())
Like save-code, but also restricts the definition to the context indicated by
.Ar t .
If
.Ar t
is \(dqs\(dq, the whole context must match; if it is \(dqe\(dq, only the
extension must match. No other values of
.Ar t
are permitted.
.It w (while: ??? condition body -> ???)
Executes
.Ar condition
and pops a value from the stack. If it is false, the command stops. Otherwise,
.Ar body
is executed and the command repeats. Notice that unlike
.Li if ,
.Ar condition

src/tgl.c
/* Main program for TGL (Text Generation Language). */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <ctype.h>
#include <stdarg.h>
#include <errno.h>

#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <fnmatch.h>
#ifdef _GNU_SOURCE
#include <getopt.h>
#endif /* _GNU_SOURCE */

#include "tgl.h"
#include "strings.h"
#include "interp.h"
#include "builtins/payload.h"

char* user_library_file, * current_context;
int suppress_unknown_alignment_warning;

/* BEGIN: Persistence */

/* Struct to head each register in the register persistence file. */
typedef struct persistent_register {
  time_t access_time;
  unsigned length;
} persistent_register;

/* Magic bytes at the beginning of the register persistence file. */
static byte register_persistence_magic[8] = {
  'T', 'g', 'l', 'V', sizeof(persistent_register), 0, 0, 0,
};

/* The format of the register persistence file is as follows:
 *   8 bytes: TglV<size of persistent_register> 0 0 0
 *     The magic header indicates the type of the file and the size of the
 *     persistent_register struct. Both are used on reading to make sure that
 *     the file is compatible.
 *   struct persistent_register { access_time = 1, length = 2 }
 *     This is used on reading to ensure that the layout of persistent_register
 *     matches what the program actually uses.
 *   256 times:
 *     struct persistent_register
 *     byte[persistent_register.length]
 *       Stores the data for each of the 256 registers, in order.
 */

/* Reads persistent registers from the given file.
 *
 * Returns 1 on success, 0 on errors. If the file is valid but truncated, the
 * registers that could be read will have been altered.
 *
 * It is not an error if the file does not exist.
 */
static int read_persistent_registers(interpreter* interp, char* filename) {
  byte magic[sizeof(register_persistence_magic)];
  persistent_register header;
  string s;
  FILE* file;
  unsigned i;

  file = fopen(filename, "rb");
  if (!file) {
    if (errno == ENOENT) {
      /* File does not exist, not an error to us. */
      return 1;
    } else {
      fprintf(stderr, "tgl: error reading register persistence file: %s\n",
              strerror(errno));
      return 0;
    }
  }

  /* Read and check magic */
  if (!fread(magic, sizeof(magic), 1, file)) {
    fprintf(stderr, "tgl: error reading register persistence file: %s\n",
            strerror(errno));
    fclose(file);
    return 0;
  }

  if (memcmp(magic, register_persistence_magic, sizeof(magic))) {
    fclose(file);
    fprintf(stderr, "tgl: register persistence file %s is incompatible\n",
            filename);
    return 0;
  }

  /* Read and check format */
  if (!fread(&header, sizeof(header), 1, file)) {
    fprintf(stderr, "tgl: error reading register persistence file: %s\n",
            strerror(errno));
    fclose(file);
    return 0;
  }

  if (header.access_time != 1 || header.length != 2) {
    fclose(file);
    fprintf(stderr, "tgl: register persistence file %s is incompatible\n",
            filename);
    return 0;
  }

  /* Read each register */
  for (i = 0; i < 256; ++i) {
    if (!fread(&header, sizeof(header), 1, file)) {
      fprintf(stderr, "tgl: register persistence file truncated\n");
      fclose(file);
      return 0;
    }

    /* Try to allocate the string, but handle allocation failure gracefully.
     * (It is conceivable we could encounter a doctored file with a really
     * large length.)
     */
    s = malloc(sizeof(struct string) + header.length);
    if (!s) {
      fprintf(stderr,
              "tgl: memory allocation for persistent register failed\n");
      fclose(file);
      return 0;
    }

    /* Set the length and read the payload in */
    s->len = header.length;
    if (s->len > 0) {
      if (s->len != fread(string_data(s), 1, s->len, file)) {
        fprintf(stderr, "tgl: register persistence file truncated\n");
        free(s);
        fclose(file);
        return 0;
      }
    }

    /* Save the register */
    free(interp->registers[i]);
    interp->registers[i] = s;
    interp->reg_access[i] = header.access_time;
  }

  /* Successful */
  fclose(file);
  return 1;
}

/* Writes persistent registers to the given file.
 *
 * Returns 1 on success, 0 on error.
 */
static int write_persistent_registers(interpreter* interp, char* filename) {
  persistent_register header;
  FILE* file;
  unsigned i;

  file = fopen(filename, "w");
  if (!file) goto error;

  if (!fwrite(register_persistence_magic,
              sizeof(register_persistence_magic), 1, file))
    goto error;

  /* Set all bytes to zero first so Valgrind won't complain.
   * (On AMD64, there is unused space between the fields which is
   * "uninitialised", though its contents do not matter to us, even on-disk.)
   */
  memset(&header, 0, sizeof(header));
  header.access_time = 1;
  header.length = 2;
  if (!fwrite(&header, sizeof(header), 1, file)) goto error;

  for (i = 0; i < 256; ++i) {
    header.access_time = interp->reg_access[i];
    header.length = interp->registers[i]->len;
    if (!fwrite(&header, sizeof(header), 1, file)) goto error;
    if (interp->registers[i]->len > 0)
      if (interp->registers[i]->len !=
          fwrite(string_data(interp->registers[i]), 1,
                 interp->registers[i]->len, file))
        goto error;
  }

  /* Success */
  fclose(file);
  return 1;

  error:
  fprintf(stderr, "tgl: error writing register persistence file: %s\n",
          strerror(errno));
  if (file)
    fclose(file);
  return 0;
}

/* END: Persistence */

/* Reads all text from the given file, then executes it.
 *
 * If scan_initial_whitespace is non-zero, the leading whitespace characters
 * are accumulated and stored in interpreter::initial_whitespace.
 *
 * If enable_history is non-zero, the interpreter's enable_history attribute
 * will be set and honoured when it finishes successfully.
 *
 * If set_global_code is true, the global_code of the payload data is set to
 * the executing code, and reset to NULL on return.
 *
 * If prefix_payload is true, try to extract prefix payload data if it can be
 * found, before executing.
 *
 * Returns an exit code.
 */
static int exec_file(interpreter* interp, FILE* file,
                     int scan_initial_whitespace,
                     int enable_history,
                     int set_global_code,
                     int prefix_payload) {
  string input;
  char buffer[1024];
  unsigned len, i;
  int status = 0;

  interp->enable_history = enable_history;

  input = empty_string();
  while (!feof(file) && !ferror(file)) {
    len = fread(buffer, 1, sizeof(buffer), file);
    input = append_data(input, buffer, buffer+len);
  }

  if (ferror(file)) {
    perror("fread");
    free(input);
    return EXIT_IO_ERROR;
  }

  if (scan_initial_whitespace) {
    for (i=0; i < input->len && isspace(string_data(input)[i]); ++i);
    if (interp->initial_whitespace)
      free(interp->initial_whitespace);
    interp->initial_whitespace = create_string(string_data(input),
                                               string_data(input)+i);
  }

  if (prefix_payload)
    input = payload_extract_prefix(input, interp);

  if (set_global_code)
    interp->payload.global_code = input;

  if (!exec_code(interp, input))
    status = EXIT_PROGRAM_ERROR;

  if (set_global_code)
    interp->payload.global_code = NULL;

  /* Add to history if appropriate */
  if (enable_history && interp->enable_history && status == 0) {
    /* Exclude the command sequence "hX" from history. */
    for (i = 0; i < input->len && isspace(string_data(input)[i]); ++i);
    if (i < input->len && 'h' == string_data(input)[i]) {
      for (++i; i < input->len && isspace(string_data(input)[i]); ++i);
      if (i < input->len && 'X' == string_data(input)[i]) {
        for (++i; i < input->len && isspace(string_data(input)[i]); ++i);
        if (i == input->len) enable_history = 0;
      }
    }

    if (enable_history) {
      /* Shift registers 0..30 back */
      free(interp->registers[0x1F]);
      memmove(interp->registers+1, interp->registers, 0x1F*sizeof(string));
      memmove(interp->reg_access+1, interp->reg_access, 0x1F*sizeof(time_t));
      /* Log new history */
      interp->registers[0] = dupe_string(input);
      touch_reg(interp, 0);
    }
  }

  free(input);
  return status;
}

/* Opens and executes the user library, then clears the stack. */
static void load_user_library(interpreter* interp) {
  FILE* file;
  int status;

  file = fopen(user_library_file, "r");
  if (!file) {
    /* If the file doesn't exist, ignore silently; otherwise, print a
     * diagnostic.
     */
    if (errno != ENOENT)
      fprintf(stderr, "tgl: unable to open user library: %s\n",
              strerror(errno));
    return;
  }

  status = exec_file(interp, file, 0, 0, 0, 0);
  fclose(file);

  /* Clear the stack and reset history offset */
  while (interp->stack)
    free(stack_pop(interp));
  interp->history_offset = 0;

  /* Print notice about error in the user library if any occurred */
  if (status)
    fprintf(stderr, "tgl: error occurred in user library\n");
}

static void print_usage(void) {
  printf("Usage: tgl [options] [infile]\nText Generation Language\n\n");
#ifdef _GNU_SOURCE
  printf(
"  -l, --library file               Use the given file (instead of ~/.tgl)\n"
"                                   for the user library.\n"
"  -r, --register-persistence file  Use the given file (instead of\n"
"                                   ~/.tgl_registers) to preserve registers.\n"
"  -c, --context name               Specify the current context.\n"
"  -p, --prefix-payload             Look for payload at the beginning of code\n"
/* -A doesn't need to be shown here. */
"  -h, --help                       This help message.\n"
    );
#else
  printf(
"  -l file  Use the given file (instead of ~/.tgl) for the user library.\n"
"  -r file  Use the given file (instead of ~/.tgl_registers) to preserve\n"
"           registers.\n"
"  -c name  Specify the current context.\n"
"  -p       Look for payload at beginning of code\n"
/* -A doesn't need to be shown here. */
"  -h       This help message.\n"
    );
#endif /* _GNU_SOURCE */
}

int main(int argc, char** argv) {
  interpreter interp;
  char reg_persistence_file_default[256];
  char user_library_file_default[256];
  char* reg_persistence_file;
  int ret, cmdstat, prefix_payload = 0;
  FILE* input;
  static char short_options[] = "l:r:c:Aph";
#ifdef _GNU_SOURCE
  static struct option long_options[] = {
   { "library", 1, NULL, 'l' },
   { "register-persistence", 1, NULL, 'r' },
   { "context", 1, NULL, 'c' },
   { "suppress-alignment-warning", 0, NULL, 'A' },
   { "prefix-payload", 0, NULL, 'p' },
   { "help", 0, NULL, 'h' },
   {0},
  };
#endif /* _GNU_SOURCE */

  /* Init default file names and options */
  snprintf(reg_persistence_file_default,
           sizeof(reg_persistence_file_default),
           "%s/.tgl_registers",
           getenv("HOME"));
  snprintf(user_library_file_default,
           sizeof(user_library_file_default),
           "%s/.tgl",
           getenv("HOME"));
  user_library_file = user_library_file_default;
  reg_persistence_file = reg_persistence_file_default;
  current_context = "";
  input = stdin;

  /* Parse command-line arguments */
  do {
#ifdef _GNU_SOURCE
    cmdstat = getopt_long(argc, argv, short_options, long_options, NULL);
#else
    cmdstat = getopt(argc, argv, short_options);
#endif /* not _GNU_SOURCE */
    switch (cmdstat) {
    case -1: break;
    case 'h':
    case '?':
      print_usage();
      return EXIT_HELP;

    case 'l':
      user_library_file = optarg;
      break;

    case 'r':
      reg_persistence_file = optarg;
      break;

    case 'c':
      current_context = optarg;
      break;

    case 'A':
      suppress_unknown_alignment_warning = 1;
      break;

    case 'p':
      prefix_payload = 1;
      break;
    }
  } while (cmdstat != -1);

  /* There should be one or zero args left. */
  if (argc - optind > 1) {
    fprintf(stderr, "tgl: too many arguments\n");
    print_usage();
    return EXIT_HELP;
  }

  if (argc - optind) {
    input = fopen(argv[optind], "r");
    if (!input) {
      fprintf(stderr, "tgl: unable to open %s: %s\n",
              argv[optind], strerror(errno));
      return EXIT_IO_ERROR;
    }
  }

  srand(time(NULL));
  interp_init(&interp);

  /* Read persistent registers */
  read_persistent_registers(&interp, reg_persistence_file);
  /* Try to execute the user library */
  load_user_library(&interp);
  /* Execute primary input */
  ret = exec_file(&interp, input, 1, 1, 1, prefix_payload);
  /* If successful, save registers */
  if (ret == 0)
    write_persistent_registers(&interp, reg_persistence_file);
  /* Done, return status to the OS */
  interp_destroy(&interp);
  return ret;
}
